<img src="../Data/images/ZumiHeader.png" width=700>

# A robot érzelmei

<font size =3> Züminek személyisége van! Ebben a leckében megtudhatod, hogyan érzékeli Zümi az emberi érzelmeket, valamint hogyan programozhatod be Zümi személyiségét. Tanulni fogsz a hangról, annak méréséről és az érzelmekhez való kapcsolódásáról. Végül megtanítod Zümit, hogy ismerje fel és reagáljon a kedvenc színedre. </font> 



## Hogyan érzékeljük az érzelmeket?
<font size =3> Vess egy pillantást az alábbi képekre, és nézd meg, hogy felismered-e az egyes érzelmeket. 

<img src="../Data/images/emotions.png" width=700> <br>

Hogyan határoztad meg, melyik érzelem melyik volt? Számos olyan jellemző lehet, mint a szem, a száj, a szemöldök és még talán a gesztusok is. Hogyan fordíthatjuk le az emberi érzelmeket robot nyelvre?

Ha láttad a *Verdák* című filmet, akkor tudhatod, hogy mindegyik autó személyiséggel rendelkezik. Hogy volt képes minden autó kifejezni az érzelmeiket? Mozdulatokkal történt? Hangokkal? Szemeikkel?</font> 


### Könyvtárak importálása
<font size =3> A személyiség függvények használatához importálnunk kell a Zümi, a képernyő és a személyiség könyvtárakat. </font> 


In [ ]:
from zumi.zumi import Zumi
from zumi.util.camera import Camera
from zumi.util.screen import Screen
from zumi.personality import Personality
from zumi.util.color_classifier import ColorClassifier
import time


zumi = Zumi()
screen = Screen()
personality = Personality(zumi, screen)
camera = Camera()

Starting Zumi 
Pi Zero I2C is available
Verified Pi Zero is the same
Gyroscope previously calibrated
Zumi board detected
Compass detected
OLED Screen detected
Gyroscope & Accelerometer detected


###  Személyiség függvények hívása
<font size =3> Íme néhány függvény, amelyeket hívhatsz:

* happy()
     
* celebrate()
       
* angry()
       
* look_around()

* look_around_open()
       
* disoriented_left()
       
* disoriented_right()

* awake()

Például a, 
<font face="Courier">personality.happy()</font> függvény Zümit ide-oda mozgatja és hangot fog kiadni!
            
Az alábbi cellában próbálj ki néhány személyiség függvényt, hogy lásd, mit csinálnak. </font>

In [ ]:
# ITT Teszteld a személyiség függvényeket!

## Hangok

<font size =3>Zümi le tud játszani olyan hangokat, amelyek megfelelnek az érzelmeinek. A hang frekvenciában és amplitúdóban mérhető. 

*   A frekvencia az impulzusok vagy rezgések száma másodpercenként, és hertz-ben mérik. Minél nagyobb a frekvencia, annál nagyobb a hangmagasság. 
*   Az amplitúdó az azt méri, mennyi hangos vagy erős a hang, és decibelben mérik. Minél nagyobb az amplitúdó, annál hangosabb a hang.

Videó: [Sound: Wavelength, Frequency, and Amplitude](https://www.youtube.com/watch?v=TsQL-sXZOLc)

Hogyan hangzanak az egyes érzelmek? A boldogság alacsony vagy magas frekvencia? A dühös alacsony vagy nagy amplitúdójú? Hogyan vonatkozik ez Zümire?

A <font face="Courier">play_note()</font> segítségével különböző hangokat játszhatsz le. Az első paraméter a lejátszani kívánt hang (C2-től B6-ig). A második paraméter opcionális, és azt az időt jelöli miliszekundumban, ameddig a hangot le szeretnéd játszani. Az alapértelmezett érték 500 ms, de ezt megváltoztathatod egy ehhez hasonló második paraméter hozzáadásával: <br><br>
<font face="Courier">play_note(Note.GS3, 400)</font>. <br><br>
Ez a Gisz hangot játssza le közép C alatt 400 milliszekundumig. Próbáld ki az alábbi kódot a skála meghallgatásához, majd írj egy saját zenét!</font>


In [ ]:
from zumi.protocol import Note 
zumi.play_note(Note.C4)
zumi.play_note(Note.D4)
zumi.play_note(Note.E4)
zumi.play_note(Note.F4)
zumi.play_note(Note.G4)
zumi.play_note(Note.A4)
zumi.play_note(Note.B4)
zumi.play_note(Note.C5)

<font size =3> Kódold saját hangjaid boldog, szomorú, dühös vagy izgatott hangokra. Próbáld ki a különböző dallamokat, amíg meg nem találod kedvenceidet. </font>

In [ ]:
# ITT Készítsd el saját dallamaidat 🎵 

## Képernyő

<font size =3> A Zümi személyiség az **OLED** (organic LED) képernyőt is használja az érzelmek megjelenítésére.
Sokféle "szeme" van Züminek:

* <font face="Courier"> close_eyes()</font>
* <font face="Courier"> sleepy_eyes()</font>
* <font face="Courier"> sleepy_left()</font>
* <font face="Courier"> sleepy_right()</font>
* <font face="Courier"> blink()</font>
* <font face="Courier"> look_around_open()</font>
* <font face="Courier"> sleeping()</font>
* <font face="Courier"> look_around()</font>     
* <font face="Courier"> glimmer()</font>
* <font face="Courier"> sad()</font>
* <font face="Courier"> happy()</font>
* <font face="Courier"> hello()</font>
* <font face="Courier"> angry()</font>

A képernyő használatához hívd a képernyőt egy általad választott függvénnyel. Próbáld ki ezt: </font>

In [ ]:
screen.sad()

### Szöveg kiírása

<font size =3> A Zümi szemeken kívül Zümi üzeneteket is megjeleníthet a képernyőjén! A <font face="Courier">draw_text()</font> függvény segítségével ilyen üzenetet is írhatsz: </font>

In [ ]:
screen.draw_text("hello!")

<font size =3> Ha azt szeretnéd, hogy a szöveg automatikusan középre kerüljön a képernyőn, hívd inkább ezt a funkciót: </font>

In [ ]:
screen.draw_text_center("hello!")

<font size =3> Ha számokat szeretnél írni, meg kell bizonyosodnod arról, hogy minden a <font face="Courier">String</font> adattípusból van. </font>

In [ ]:
number = 10
screen.draw_text("ten " + str(number)) # a str() függvény átalakítja a számot karakterré

<font size =3> Akár arra is megkérheted Zümit, hogy jelenítse meg neked az időt! </font>

In [ ]:
for i in range(0,50):
    screen.draw_text_center(time.ctime())
    time.sleep(0.1)

## Rajtad tartja a szemét
<font size = 3> Ebben a tevékenységben Zümi rajtad tartja a szemét. Egy kódot fogsz írni, hogy Zümi szemei ​​kövessék az arcodat, amikor a látótávolságának középső, bal vagy jobb oldalán van. Ha Zümi nem lát téged, szomorú lesz!
Ez a tevékenység a 4.2 lecke Arcfelismerés kódját használja. Vess egy pillantást arra a leckére, ha még nem tetted meg! </font>
    
### A keret felosztása balra, jobbra és középre
<font size=3> Az első lépés annak kiderítése, hogy hány pixel széles a Zümi kameraképe. Ehhez a kameraobjektum tulajdonságait, így érhetjük el:

In [ ]:
print(camera.image_width)

160


<font size=3> Alapértelmezés szerint a pixel szélessége mindig 160 lesz. Bár ezen változtathatsz, a példánkban ezt az értéket fogjuk használni. Oszd fel ezt a képkeretet olyan területekre, amelyek neked a keret bal oldalát, jobb oldalát és a közepét jelentik. Szétoszthatod egyenlően, vagy sem. Tőled függ! Ne feledd, hogy 160 pixel van három részre osztva, és nem lehet fél pixelt használni.

## Ha te mosolyogsz, én is mosolygok

<font size=3> Most, hogy megtanultad használni a hangokat és a képernyőt, hogy Züminek személyiséget adhass, tanítsd meg neki, hogy reagáljon a mosolyra! Egy előző leckében már valószínűleg megtanultad a <font face="Courier">vision.find_face()</font> függvény arcok keresésére való használatát. Mit gondolsz, milyen jellemzőkkel bír a mosoly? A szánk sötétebb képpontjai felfelé vagy lefelé vannak döntve? Ebben a leckében használni fogsz egy függvényt, hogy ellenőrizd, hogy az arc, amit Zümi észlel, mosolyog-e. Ha meglát egy mosolyt, akkor boldog lesz. Ha nem, adj Züminek szomorú szemeket.</font>
    
### vision.find_smile()
<font size=3> A függvény, amelyet erre a tevékenységre használsz a: <font face="Courier">vision.find_smile()</font>. Ez a függvény visszaadja a pixelkoordinátákat, ha Zümi mosolyt lát, vagy <font face="Courier">(-1,-1,-1,-1)</font> ha nem. Például, ha képet készítesz, és képként mented, ellenőrizheted, hogy Zümi talált-e mosolyt vagy sem: <br> <br>
    
<font face="Courier">
    
(x,y,w,h) = vision.find_smile(image)<br>
if (x,y,w,h) == (-1,-1,-1,-1): <br>
<span style="margin-left: 40px;"># Zumi did not find a smile :(</span> <br>
else: <br>
<span style="margin-left: 40px;"># Zumi sees a smile! :)</span> <br>
    </font> </font>
    
### Kódold!
<font size=3> Most írj egy kódot, amely Zümi kamerájával képet készít, és mosolyt keres. **Tipp:** Ne felejtsd el megváltoztatni a paramétereket, ha Zümi nem talál mosolyt. Adj Züminek személyiséget, és párosítsd érzelmeit a tieiddel! Töltsd ki az alábbi kódsablont:

In [ ]:
camera.start_camera()
try:
    for i in range(1000):
        # Fejezd be a kódot!

        
finally:
    camera.close()

<hr>

# Extra Tevékenységek <br> 

<img src="../Data/images/physics_extension.jpg" width=70 align="left">

###  Frekvencia <br> 
<font size =3>Hangolók segítségével azonosítsd a különböző Zümi hangok frekvenciáját, hullámhosszát és amplitúdóját. Melyik érzelemnek van a legnagyobb vagy a legkisebb frekvenciája, hullámhossza és amplitúdója? </font><br><br>


### Adj személyiséget a színosztályozó órához!
<font size =3> <span style="color:red"> **Megjegyzés!** </span> Ehhez a tevékenységhez a "Felfedezés" oldalon található Color Training Wizard szükséges.<br><br>
Menj vissza a Color Training Wizard-hoz, és tanítsd meg Zümit többféle színre. Töltsd be az alábbi modellt, és adj boldog reakciót a kedvenc színéhez, és szomorú vagy dühös reakciók a többi színre. Kérd meg egy társad, hogy mutasson Züminek különféle színeket, és találd ki, mi Zümi kedvenc színe! </font> 

In [ ]:
camera = Camera()
knn = ColorClassifier()
train = knn.load_model("PROJECT NAME HERE")
knn.fit("hsv")

camera.start_camera()

while True:
        user_input = input("Press 'enter' to predict or 'q to quit: ")

        if user_input == "q":
            break
            
        image = camera.capture()
        predict = knn.predict(image)
        screen.draw_text_center(predict)
        
        # IDE Add hozzá az if állításokat!

camera.close()